packages/fluentui/react-northstar/src/components/Dropdown/Dropdown.tsx

```diff
@@ -826,31 +817,154 @@ class Dropdown extends AutoControlledComponent<WithAsProp<DropdownProps>, Dropdo
+          const shouldMenuClose = newSearchQuery === '' || newValue !== undefined;
+
+          if (shouldMenuClose) {
+            newState.open = false;
+          }
+        } else {
+          newState.open = true;
         }
-      } else {
-        newState.highlightedIndex = null;
+
+        break;
       }
+      case Downshift.stateChangeTypes.keyDownEnter:
+      case Downshift.stateChangeTypes.clickItem:
+        const shouldAddHighlightedIndex = !multiple && !!items;
+
+        newState.searchQuery = this.getSelectedItemAsString(newValue);
+        newState.value = multiple ? [...value, newValue] : [newValue];
+        newState.open = false;
+
+        if (shouldAddHighlightedIndex) {
+          newState.highlightedIndex = items.indexOf(newValue);
+        }
+
+        if (getA11ySelectionMessage && getA11ySelectionMessage.onAdd) {
+          this.setA11ySelectionMessage(getA11ySelectionMessage.onAdd(newValue));
+        }
+
+        if (multiple) {
+          setTimeout(() => (this.selectedItemsRef.current.scrollTop = this.selectedItemsRef.current.scrollHeight), 0);
+        }
+
+        this.tryFocusTriggerButton();
+
+        break;
+      case Downshift.stateChangeTypes.keyDownEscape:
+        if (search) {
+          newState.searchQuery = '';
+
+          if (!multiple) {
+            newState.value = [];
+          }
+        }
+        newState.open = false;
+        newState.highlightedIndex = highlightFirstItemOnOpen ? 0 : null;
+        break;
+      case Downshift.stateChangeTypes.keyDownArrowDown:
+      case Downshift.stateChangeTypes.keyDownArrowUp:
+        if (newOpen !== undefined) {
+          newState.open = newOpen;
+          newState.highlightedIndex = newHighlightedIndex;
+
+          if (newOpen) {
+            const highlightedIndexOnArrowKeyOpen = this.getHighlightedIndexOnArrowKeyOpen(changes);
+
+            if (_.isNumber(highlightedIndexOnArrowKeyOpen)) {
+              newState.highlightedIndex = highlightedIndexOnArrowKeyOpen;
+            }
+
+            if (!search) {
+              this.listRef.current.focus();
+            }
+          } else {
+            newState.highlightedIndex = null;
+          }
+        }
+      case Downshift.stateChangeTypes['keyDownHome']:
+      case Downshift.stateChangeTypes['keyDownEnd']:
+        if (open && _.isNumber(newHighlightedIndex)) {
+          newState.highlightedIndex = newHighlightedIndex;
+          newState.itemIsFromKeyboard = true;
+        }
 
-      this.setStateAndInvokeHandler(['onOpenChange'], null, newState);
+        break;
+      case Downshift.stateChangeTypes.clickButton:
+      case Downshift.stateChangeTypes.keyDownSpaceButton:
+        newState.open = newOpen;
+
+        if (newOpen) {
+          const highlightedIndexOnArrowKeyOpen = this.getHighlightedIndexOnArrowKeyOpen(changes);
+
+          if (_.isNumber(highlightedIndexOnArrowKeyOpen)) {
+            newState.highlightedIndex = highlightedIndexOnArrowKeyOpen;
+          }
```

Comment:
As I `getHighlightedIndexOnArrowKeyOpen()` is used twice in current code, I suggest to modify to provide proper results:

```diff
-  getHighlightedIndexOnArrowKeyOpen = (changes: StateChangeOptions<ShorthandValue<DropdownItemProps>>): number => {
+  getHighlightedIndexOnArrowKeyOpen = (changes: StateChangeOptions<ShorthandValue<DropdownItemProps>>): number | undefined => {
    const { filteredItems, highlightedIndex, value } = this.state;
    const { highlightFirstItemOnOpen, items, multiple, search } = this.props;
    const isArrowUp = changes.type === Downshift.stateChangeTypes.keyDownArrowUp;
    const isArrowDown = changes.type === Downshift.stateChangeTypes.keyDownArrowDown;
    const itemsLength = filteredItems.length;

    if (highlightedIndex) {
      return highlightedIndex;
    }

    if (highlightFirstItemOnOpen) {
      // otherwise, if highlightFirstItemOnOpen prop is provied, highlight first item.
      return 0;
    }

    if (!multiple && !search && value.length > 0) {
      // in single selection, if there is a selected item, highlight it.
      const offset = isArrowUp ? -1 : isArrowDown ? 1 : 0;
      const newHighlightedIndex = items.indexOf(value[0]) + offset;
      if (newHighlightedIndex >= itemsLength) {
        return 0;
      }
      if (newHighlightedIndex < 0) {
        return itemsLength - 1;
      }
      return newHighlightedIndex;
    }

    if (isArrowDown) {
      return 0;
    }
    if (isArrowUp) {
      return itemsLength - 1;
    }

-    return null;
+    return undefined;
  };
```

```suggestion
          newState.highlightedIndex = this.getHighlightedIndexOnArrowKeyOpen(changes);
```

Reply:
Another option is to integrate the check for `newOpen` to `getHighlightedIndexOnArrowKeyOpen` as it already accepts `changes`